compiler/hash-ast-passes/src/analysis/mod.rs

```diff
@@ -32,11 +34,17 @@ pub struct SemanticAnalyser<'s> {
     /// The current scope of the traversal, representing which block the
     /// analyser is walking.
     pub(crate) current_block: BlockOrigin,
+    /// The kind of module that the semantic analysis is currently traversing.
+    pub(crate) module_kind: Option<ModuleKind>,
 }
 
 impl<'s> SemanticAnalyser<'s> {
     /// Create a new semantic analyser
-    pub fn new(source_map: &'s SourceMap, source_id: SourceId) -> Self {
+    pub fn new(
+        source_map: &'s SourceMap,
+        source_id: SourceId,
+        module_kind: Option<ModuleKind>,
```

Comment:
SourceMap should have a field that maps ModuleId to ModuleKind, instead of having to pass in this parameter. And then this can be looked up based on source_id.